File: scripts/cron.py
#!/usr/bin/env python3
"""Companion scheduled tasks — launchd control plane.

Usage:
    python scripts/cron.py list                     # Show all companion jobs
    python scripts/cron.py add <name> <cron> <cmd>  # Add a job
    python scripts/cron.py remove <name>            # Remove a job
    python scripts/cron.py edit <name> <cron>       # Change schedule
    python scripts/cron.py run <name>               # Run a job now (manual trigger)
    python scripts/cron.py install                  # Install all jobs to launchd
    python scripts/cron.py uninstall                # Uninstall all jobs from launchd

Jobs are defined in a YAML-like config at scripts/jobs.json.
Plists are generated and managed in ~/Library/LaunchAgents/.
"""
import argparse
import json
import os
import plistlib
import subprocess
import sys
from pathlib import Path

PROJECT_ROOT = Path(__file__).parent.parent
AGENT_NAME = os.environ.get("AGENT", "companion")
JOBS_FILE = PROJECT_ROOT / "scripts" / "agents" / AGENT_NAME / "jobs.json"
LAUNCH_AGENTS = Path.home() / "Library" / "LaunchAgents"
LABEL_PREFIX = f"com.atrophiedmind.{AGENT_NAME}."
PYTHON = sys.executable
LOGS_DIR = PROJECT_ROOT / "logs" / AGENT_NAME


def _load_jobs() -> dict:
    if JOBS_FILE.exists():
        return json.loads(JOBS_FILE.read_text())
    return {}


def _save_jobs(jobs: dict):
    JOBS_FILE.write_text(json.dumps(jobs, indent=2))


def _parse_cron(cron_str: str) -> dict:
    """Parse '17 3 * * *' into launchd StartCalendarInterval dict."""
    parts = cron_str.split()
    if len(parts) != 5:
        raise ValueError(f"Invalid cron: '{cron_str}' — need 5 fields: min hour dom month dow")

    minute, hour, dom, month, dow = parts
    interval = {}

    if minute != '*':
        interval['Minute'] = int(minute)
    if hour != '*':
        interval['Hour'] = int(hour)
    if dom != '*':
        interval['Day'] = int(dom)
    if month != '*':
        interval['Month'] = int(month)
    if dow != '*':
        interval['Weekday'] = int(dow)

    return interval


def _plist_path(name: str) -> Path:
    return LAUNCH_AGENTS / f"{LABEL_PREFIX}{name}.plist"


def _generate_plist(name: str, job: dict) -> dict:
    """Generate a launchd plist dict for a job.

    Supports two types:
      - "calendar" (default): uses StartCalendarInterval with a cron string
      - "interval": uses StartInterval with seconds-based scheduling
    """
    label = f"{LABEL_PREFIX}{name}"
    script_path = str(PROJECT_ROOT / job["script"])

    log_path = str(LOGS_DIR / f"{name}.log")

    extra_args = job.get("args", [])
    plist = {
        'Label': label,
        'ProgramArguments': [PYTHON, script_path] + extra_args,
        'WorkingDirectory': str(PROJECT_ROOT),
        'StandardOutPath': log_path,
        'StandardErrorPath': log_path,
        'EnvironmentVariables': {
            'PATH': f"/usr/local/bin:/usr/bin:/bin:{Path(PYTHON).parent}",
            'AGENT': AGENT_NAME,
        },
    }

    job_type = job.get("type", "calendar")
    if job_type == "interval":
        plist['StartInterval'] = int(job["interval_seconds"])
    else:
        plist['StartCalendarInterval'] = _parse_cron(job["cron"])

    return plist


def cmd_list(args):
    jobs = _load_jobs()
    if not jobs:
        print("  No jobs configured.")
        return

    print()
    print(f"  {'Name':<20} {'Schedule':<20} {'Script':<35} {'Installed'}")
    print(f"  {'─'*20} {'─'*20} {'─'*35} {'─'*10}")

    for name, job in jobs.items():
        installed = "yes" if _plist_path(name).exists() else "no"
        if job.get("type") == "interval":
            schedule = f"every {job['interval_seconds']}s"
        else:
            schedule = job.get("cron", "?")
        print(f"  {name:<20} {schedule:<20} {job['script']:<35} {installed}")
    print()


def cmd_add(args):
    jobs = _load_jobs()
    name = args.name
    cron = args.cron
    script = args.script

    # Validate cron
    try:
        _parse_cron(cron)
    except ValueError as e:
        print(f"  Error: {e}")
        return

    # Validate script exists
    script_path = PROJECT_ROOT / script
    if not script_path.exists():
        print(f"  Warning: Script not found at {script_path}")

    jobs[name] = {
        "cron": cron,
        "script": script,
        "description": args.description or "",
    }
    _save_jobs(jobs)
    print(f"  Added job '{name}': {cron} → {script}")

    if args.install:
        _install_job(name, jobs[name])


def cmd_remove(args):
    jobs = _load_jobs()
    name = args.name

    if name not in jobs:
        print(f"  Job '{name}' not found.")
        return

    # Uninstall first
    _uninstall_job(name)

    del jobs[name]
    _save_jobs(jobs)
    print(f"  Removed job '{name}'.")


def cmd_edit(args):
    jobs = _load_jobs()
    name = args.name

    if name not in jobs:
        print(f"  Job '{name}' not found.")
        return

    try:
        _parse_cron(args.cron)
    except ValueError as e:
        print(f"  Error: {e}")
        return

    old_cron = jobs[name]["cron"]
    jobs[name]["cron"] = args.cron
    _save_jobs(jobs)
    print(f"  Updated '{name}': {old_cron} → {args.cron}")

    # Reinstall if already installed
    if _plist_path(name).exists():
        _uninstall_job(name)
        _install_job(name, jobs[name])
        print(f"  Reinstalled with new schedule.")


def cmd_run(args):
    jobs = _load_jobs()
    name = args.name

    if name not in jobs:
        print(f"  Job '{name}' not found.")
        return

    job = jobs[name]
    script_path = PROJECT_ROOT / job["script"]
    extra_args = job.get("args", [])
    print(f"  Running {script_path} {' '.join(extra_args)}...")
    result = subprocess.run(
        [PYTHON, str(script_path)] + extra_args,
        cwd=str(PROJECT_ROOT),
    )
    sys.exit(result.returncode)


def _install_job(name: str, job: dict):
    LOGS_DIR.mkdir(exist_ok=True)
    plist = _generate_plist(name, job)
    plist_path = _plist_path(name)

    with open(plist_path, 'wb') as f:
        plistlib.dump(plist, f)

    subprocess.run(["launchctl", "load", str(plist_path)], check=False)
    print(f"  Installed {name} → {plist_path}")


def _uninstall_job(name: str):
    plist_path = _plist_path(name)
    if plist_path.exists():
        subprocess.run(["launchctl", "unload", str(plist_path)], check=False)
        plist_path.unlink()
        print(f"  Uninstalled {name}")


def cmd_install(args):
    jobs = _load_jobs()
    if not jobs:
        print("  No jobs to install.")
        return

    LOGS_DIR.mkdir(exist_ok=True)
    for name, job in jobs.items():
        _install_job(name, job)
    print(f"\n  Installed {len(jobs)} job(s).")


def cmd_uninstall(args):
    jobs = _load_jobs()
    if not jobs:
        print("  No jobs configured.")
        return

    for name in jobs:
        _uninstall_job(name)
    print(f"\n  Uninstalled {len(jobs)} job(s).")


def main():
    global AGENT_NAME, JOBS_FILE, LABEL_PREFIX, LOGS_DIR

    parser = argparse.ArgumentParser(description="Agent scheduled tasks")
    parser.add_argument("--agent", default=None, help="Agent name (default: from AGENT env var)")
    sub = parser.add_subparsers(dest="command")

    sub.add_parser("list", help="Show all jobs")

    add_p = sub.add_parser("add", help="Add a job")
    add_p.add_argument("name", help="Job name (e.g. introspect)")
    add_p.add_argument("cron", help="Cron schedule (e.g. '17 3 * * *')")
    add_p.add_argument("script", help="Script path relative to project root (e.g. scripts/introspect.py)")
    add_p.add_argument("-d", "--description", help="Job description")
    add_p.add_argument("--install", action="store_true", help="Install immediately")

    rm_p = sub.add_parser("remove", help="Remove a job")
    rm_p.add_argument("name", help="Job name")

    edit_p = sub.add_parser("edit", help="Change a job's schedule")
    edit_p.add_argument("name", help="Job name")
    edit_p.add_argument("cron", help="New cron schedule")

    run_p = sub.add_parser("run", help="Run a job now")
    run_p.add_argument("name", help="Job name")

    sub.add_parser("install", help="Install all jobs to launchd")
    sub.add_parser("uninstall", help="Uninstall all jobs from launchd")

    args = parser.parse_args()

    if args.agent:
        AGENT_NAME = args.agent
        JOBS_FILE = PROJECT_ROOT / "scripts" / "agents" / AGENT_NAME / "jobs.json"
        LABEL_PREFIX = f"com.atrophiedmind.{AGENT_NAME}."
        LOGS_DIR = PROJECT_ROOT / "logs" / AGENT_NAME

    if not args.command:
        parser.print_help()
        return

    {
        "list": cmd_list,
        "add": cmd_add,
        "remove": cmd_remove,
        "edit": cmd_edit,
        "run": cmd_run,
        "install": cmd_install,
        "uninstall": cmd_uninstall,
    }[args.command](args)


if __name__ == "__main__":
    main()

File: src/main/cron.ts
/**
 * launchd job management - scheduled task control plane.
 * Port of scripts/cron.py.
 *
 * Generates launchd plists for companion background jobs. Jobs are defined
 * in per-agent jobs.json files. Supports calendar-based (cron) and
 * interval-based scheduling.
 */

import { execSync, spawnSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import { getConfig, BUNDLE_ROOT, USER_DATA } from './config';
import { createLogger } from './logger';

const log = createLogger('cron');

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface Job {
  cron?: string;
  script: string;
  description?: string;
  args?: string[];
  type?: 'calendar' | 'interval';
  interval_seconds?: number;
}

export interface JobInfo extends Job {
  name: string;
  installed: boolean;
  schedule: string;
}

// ---------------------------------------------------------------------------
// Paths
// ---------------------------------------------------------------------------

const LAUNCH_AGENTS = path.join(process.env.HOME || '/tmp', 'Library', 'LaunchAgents');

function labelPrefix(): string {
  return `com.atrophiedmind.${getConfig().AGENT_NAME}.`;
}

function jobsFile(): string {
  const config = getConfig();
  return path.join(BUNDLE_ROOT, 'scripts', 'agents', config.AGENT_NAME, 'jobs.json');
}

function logsDir(): string {
  const config = getConfig();
  return path.join(USER_DATA, 'logs', config.AGENT_NAME);
}

function plistPath(name: string): string {
  return path.join(LAUNCH_AGENTS, `${labelPrefix()}${name}.plist`);
}

// ---------------------------------------------------------------------------
// Job storage
// ---------------------------------------------------------------------------

function loadJobs(): Record<string, Job> {
  const p = jobsFile();
  if (!fs.existsSync(p)) return {};
  try {
    return JSON.parse(fs.readFileSync(p, 'utf-8'));
  } catch {
    return {};
  }
}

function saveJobs(jobs: Record<string, Job>): void {
  const p = jobsFile();
  fs.mkdirSync(path.dirname(p), { recursive: true });
  fs.writeFileSync(p, JSON.stringify(jobs, null, 2));
}

// ---------------------------------------------------------------------------
// Cron parsing
// ---------------------------------------------------------------------------

interface CalendarInterval {
  Minute?: number;
  Hour?: number;
  Day?: number;
  Month?: number;
  Weekday?: number;
}

function parseCron(cronStr: string): CalendarInterval {
  const parts = cronStr.split(/\s+/);
  if (parts.length !== 5) {
    throw new Error(`Invalid cron: '${cronStr}' - need 5 fields: min hour dom month dow`);
  }

  const [minute, hour, dom, month, dow] = parts;
  const interval: CalendarInterval = {};

  if (minute !== '*') interval.Minute = parseInt(minute, 10);
  if (hour !== '*') interval.Hour = parseInt(hour, 10);
  if (dom !== '*') interval.Day = parseInt(dom, 10);
  if (month !== '*') interval.Month = parseInt(month, 10);
  if (dow !== '*') interval.Weekday = parseInt(dow, 10);

  return interval;
}

// ---------------------------------------------------------------------------
// Plist generation
// ---------------------------------------------------------------------------

interface PlistDict {
  Label: string;
  ProgramArguments: string[];
  WorkingDirectory: string;
  StandardOutPath: string;
  StandardErrorPath: string;
  EnvironmentVariables: Record<string, string>;
  StartCalendarInterval?: CalendarInterval;
  StartInterval?: number;
}

function generatePlist(name: string, job: Job): PlistDict {
  const config = getConfig();
  const label = `${labelPrefix()}${name}`;
  const scriptPath = path.resolve(BUNDLE_ROOT, job.script);
  const logPath = path.join(logsDir(), `${name}.log`);
  const extraArgs = job.args || [];

  const plist: PlistDict = {
    Label: label,
    ProgramArguments: [config.PYTHON_PATH, scriptPath, ...extraArgs],
    WorkingDirectory: BUNDLE_ROOT,
    StandardOutPath: logPath,
    StandardErrorPath: logPath,
    EnvironmentVariables: {
      PATH: `/usr/local/bin:/usr/bin:/bin:${path.dirname(config.PYTHON_PATH)}`,
      AGENT: config.AGENT_NAME,
    },
  };

  const jobType = job.type || 'calendar';
  if (jobType === 'interval' && job.interval_seconds) {
    plist.StartInterval = job.interval_seconds;
  } else if (job.cron) {
    plist.StartCalendarInterval = parseCron(job.cron);
  }

  return plist;
}

// Serialize plist to XML (minimal implementation)
function plistToXml(plist: PlistDict): string {
  const lines: string[] = [
    '<?xml version="1.0" encoding="UTF-8"?>',
    '<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">',
    '<plist version="1.0">',
    '<dict>',
  ];

  function escapeXml(s: string): string {
    return s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;');
  }

  function addKeyValue(key: string, value: unknown, indent = 1): void {
    const tab = '\t'.repeat(indent);
    if (typeof value === 'string') {
      lines.push(`${tab}<key>${escapeXml(key)}</key>`);
      lines.push(`${tab}<string>${escapeXml(value)}</string>`);
    } else if (typeof value === 'number') {
      lines.push(`${tab}<key>${escapeXml(key)}</key>`);
      lines.push(`${tab}<integer>${value}</integer>`);
    } else if (typeof value === 'boolean') {
      lines.push(`${tab}<key>${escapeXml(key)}</key>`);
      lines.push(`${tab}<${value}/>`);
    } else if (Array.isArray(value)) {
      lines.push(`${tab}<key>${escapeXml(key)}</key>`);
      lines.push(`${tab}<array>`);
      for (const item of value) {
        if (typeof item === 'string') {
          lines.push(`${tab}\t<string>${escapeXml(item)}</string>`);
        }
      }
      lines.push(`${tab}</array>`);
    } else if (typeof value === 'object' && value !== null) {
      lines.push(`${tab}<key>${escapeXml(key)}</key>`);
      lines.push(`${tab}<dict>`);
      for (const [k, v] of Object.entries(value)) {
        addKeyValue(k, v, indent + 1);
      }
      lines.push(`${tab}</dict>`);
    }
  }

  for (const [key, value] of Object.entries(plist)) {
    addKeyValue(key, value);
  }

  lines.push('</dict>');
  lines.push('</plist>');
  return lines.join('\n');
}

// ---------------------------------------------------------------------------
// Install / Uninstall
// ---------------------------------------------------------------------------

function installJob(name: string, job: Job): void {
  fs.mkdirSync(logsDir(), { recursive: true });
  fs.mkdirSync(LAUNCH_AGENTS, { recursive: true });

  const plist = generatePlist(name, job);
  const pp = plistPath(name);

  fs.writeFileSync(pp, plistToXml(plist));
  spawnSync('launchctl', ['load', pp], { stdio: 'pipe' });
  log.info(`Installed ${name} -> ${pp}`);
}

function uninstallJob(name: string): void {
  const pp = plistPath(name);
  if (fs.existsSync(pp)) {
    spawnSync('launchctl', ['unload', pp], { stdio: 'pipe' });
    fs.unlinkSync(pp);
    log.info(`Uninstalled ${name}`);
  }
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export function listJobs(): JobInfo[] {
  const jobs = loadJobs();
  return Object.entries(jobs).map(([name, job]) => ({
    name,
    ...job,
    installed: fs.existsSync(plistPath(name)),
    schedule: job.type === 'interval'
      ? `every ${job.interval_seconds}s`
      : job.cron || '?',
  }));
}

export function addJob(
  name: string,
  cronStr: string,
  script: string,
  description = '',
  install = false,
): void {
  parseCron(cronStr); // validate
  const jobs = loadJobs();
  jobs[name] = { cron: cronStr, script, description };
  saveJobs(jobs);
  log.info(`Added job '${name}': ${cronStr} -> ${script}`);
  if (install) installJob(name, jobs[name]);
}

export function removeJob(name: string): void {
  const jobs = loadJobs();
  if (!(name in jobs)) {
    log.warn(`Job '${name}' not found`);
    return;
  }
  uninstallJob(name);
  delete jobs[name];
  saveJobs(jobs);
  log.info(`Removed job '${name}'`);
}

export function editJobSchedule(name: string, cronStr: string): void {
  const jobs = loadJobs();
  if (!(name in jobs)) {
    log.warn(`Job '${name}' not found`);
    return;
  }
  parseCron(cronStr);
  jobs[name].cron = cronStr;
  saveJobs(jobs);

  // Reinstall if already installed
  if (fs.existsSync(plistPath(name))) {
    uninstallJob(name);
    installJob(name, jobs[name]);
  }
}

// ---------------------------------------------------------------------------
// Job run history (in-memory, last 100 entries)
// ---------------------------------------------------------------------------

export interface JobRunEntry {
  name: string;
  timestamp: string;
  exitCode: number;
  output: string;
  durationMs: number;
  agent: string;
}

const _history: JobRunEntry[] = [];
const MAX_HISTORY = 100;

function pushHistory(entry: JobRunEntry): void {
  _history.push(entry);
  if (_history.length > MAX_HISTORY) _history.splice(0, _history.length - MAX_HISTORY);
}

export function getJobHistory(): JobRunEntry[] {
  return [..._history].reverse();
}

export function runJobNow(name: string): JobRunEntry {
  const jobs = loadJobs();
  const config = getConfig();
  const t0 = Date.now();

  if (!(name in jobs)) {
    log.warn(`Job '${name}' not found`);
    const entry: JobRunEntry = {
      name,
      timestamp: new Date().toISOString(),
      exitCode: 1,
      output: `Job '${name}' not found`,
      durationMs: Date.now() - t0,
      agent: config.AGENT_NAME,
    };
    pushHistory(entry);
    return entry;
  }

  const job = jobs[name];
  const scriptPath = path.resolve(BUNDLE_ROOT, job.script);
  const extraArgs = job.args || [];

  const result = spawnSync(config.PYTHON_PATH, [scriptPath, ...extraArgs], {
    cwd: BUNDLE_ROOT,
    stdio: 'pipe',
    env: { ...process.env, AGENT: config.AGENT_NAME },
    timeout: 120_000,
  });

  const stdout = result.stdout?.toString('utf-8') || '';
  const stderr = result.stderr?.toString('utf-8') || '';
  const output = (stdout + (stderr ? '\n--- stderr ---\n' + stderr : '')).trim();
  const exitCode = result.status ?? 1;

  const entry: JobRunEntry = {
    name,
    timestamp: new Date().toISOString(),
    exitCode,
    output: output.slice(0, 10_000),
    durationMs: Date.now() - t0,
    agent: config.AGENT_NAME,
  };
  pushHistory(entry);

  log.info(`Job '${name}' finished (exit=${exitCode}, ${entry.durationMs}ms)`);
  return entry;
}

/**
 * Read the last N lines of a job's log file.
 */
export function readJobLog(name: string, lines = 200): string {
  const logPath = path.join(logsDir(), `${name}.log`);
  if (!fs.existsSync(logPath)) return '';
  try {
    const content = fs.readFileSync(logPath, 'utf-8');
    const allLines = content.split('\n');
    return allLines.slice(-lines).join('\n');
  } catch {
    return '';
  }
}

export function installAllJobs(): void {
  const jobs = loadJobs();
  if (!Object.keys(jobs).length) {
    log.info('No jobs to install');
    return;
  }
  for (const [name, job] of Object.entries(jobs)) {
    installJob(name, job);
  }
  log.info(`Installed ${Object.keys(jobs).length} job(s)`);
}

export function uninstallAllJobs(): void {
  const jobs = loadJobs();
  for (const name of Object.keys(jobs)) {
    uninstallJob(name);
  }
  log.info(`Uninstalled ${Object.keys(jobs).length} job(s)`);
}

export function toggleCron(enabled: boolean): void {
  if (enabled) {
    installAllJobs();
  } else {
    uninstallAllJobs();
  }
}
